[PROPOSAL] Decoupling triggers proposal

andresmgot opened the pull request:
This is a design proposal for decoupling triggers and runtimes in different logic entities.
cc/ @deissnerk @anguslees @sebgoa @ngtuna

mike-pun commented:
The function input protocol should be agnostic to the trigger type so that we do not have to "fit it" into the behavior of a particular protocol eg. HTTP. I am looking extending the Kafka runtime to support Kafka features such as consumer groups, manual commits,etc.. some of these features are essential to a lot of use cases. I was thinking of doing a custom runtime for us, but it would be better if the protocol supports such extensions natively.

anguslees commented:
@mike-pun: Agreed.  Our rough plan atm (as described in this PR) is to pick a single RPC protocol, and rewrite all the language-runtimes to implement that, and the trigger-runtimes to invoke that.  At the moment, it looks like the RPC protocol we'll probably choose is regular HTTP, since it has extremely broad language support, and a couple of nice features as standard (arbitrary key/value metadata, self-describing request/reply body format, etc).

So: with this plan, the kafka trigger (consumer) _will_ have to turn kafka events into HTTP requests under the hood (and probably ignore the response body, provided it gets a 2xx), but it _won't_ have to pretend to be an HTTP trigger to the user (ie: the cli and Function spec will present a clearly distinct "kafka" event trigger type, that can have its own options, etc).

Does that make sense?

mikep-locol-media commented:
Oh. 

So the trigger-runtimes will actually use HTTP to communicate to the function? Ok. I will wait to see how this develops.

phsiao commented:
I am new to this proposal, so I am not aware of previous discussion and forgive me if I am just bringing up ideas that have been discussed.

When I look at this problem, _my intuition is we should implement each trigger as a sidecar container, and they don't need to be exclusive_.   For example, you can have a trigger-http container, and a trigger-event container, or even a trigger-scheduled and others.   It decouples the implementation of trigger and runtime in the most ideal way IMO.   And it looks like there is an agreement on that the runtime should implement a unified interface, be that can be make as general as possible while remain extensible.   At the same time, people are free to add/change/del triggers as well as their implementations as they see fit.

Trigger being a sidecar container also has the advantage of allowing layering and composition on top of it because it is still in the pod.   For example, you can add load balancer in front of it, you can use istio in front of it for additional policy control, and it can also be just used as a simple service that can be reached by cluster IP, and network policy can still be applied at that level.

The downside of the sidecar is there are more moving parts that can fail independently and cause the pod to be not ready, and will take longer to start, but I think these are technical challenges that we can address.

andresmgot commented:
These are the issues I see for using sidecar containers instead of standalone pods:
 - It forces to have all the possible triggers. We can assume that we want to keep adding different triggers to Kubeless (for example RabbitMQ). That will force that for every deployment the user will need to deploy Kafka and RabbitMQ in order to run the function pod, even if the user is not interested in a particular event trigger. 
 - You will need to have the same container in every pod. Instead of having just one centralized pod/container that handle a certain event source, we will have N identical containers running in N pods. AFAIK that would increase the operational cost.
 - In case we don't want to include all the triggers it requires a redeployment every time a new trigger type is added to a function (it is not possible to modify the amount of containers of a running pod).

phsiao commented:
@andresmgot I don't agree that the sidecar containers would force to have all possible triggers.   The kubeless controller still controls the pod spec, and can decide what sidecar containers to be injected into the pod.   It should not require redeployment of pod to keep up with all possible triggers.   Maybe I am missing something?

Regarding duplication of containers, yes, it is true, but these trigger-bridging sidecars should be small and take little resource, and it is a reasonable tradeoff in my opinion.

andresmgot commented:
Yes sorry, I meant that either we deploy every function with all the trigger controllers (first bullet point) or we need to redeploy the runtime when a new trigger type is added (third point). AFAIK you cannot add a new container to a running pod so it will require to do the redeployment. Isn't it?

It will also slow down the deployment a little bit as you mentioned. 

You can still put a LoadBalancer in front of it if the runtime has a single container and it will help to isolate the development/maintenance of the trigger controllers and the runtimes if they are in separated pods. I may be missing a key benefit of having all the containers in the same pod, could you elaborate?

deissnerk commented:
> For example, you can add load balancer in front of it, you can use istio in front of it for additional policy control, and it can also be just used as a simple service that can be reached by cluster IP, and network policy can still be applied at that level.

@phsiao How would you use istio in front of a trigger side car? I thought, istio comes with Envoy as its own side car. Sorry, I don't know much about istio, yet. Could you describe, what you have in mind?

phsiao commented:
`I meant that either we deploy every function with all the trigger controllers (first bullet point) or we need to redeploy the runtime when a new trigger type is added (third point). AFAIK you cannot add a new container to a running pod so it will require to do the redeployment. Isn't it?`

If I read this and the previous comment correctly, you are considering the case where a user updates a deployed function, especially by its changing/adding triggers.   It is true, with sidecar container bridging triggers, every time the trigger definition change a new pod would need to be created.  I'd argue that we should not allow adding/updating trigger to a deployed function, because it should be just as easy to spin up new ones, and it is likely preferred because the `interface` of the function is different now and we can avoid dealing with the resulting dependency resolution problem, for example, finding out how many functions still depend on the trigger-http of a function when that is about to be changed.

`I may be missing a key benefit of having all the containers in the same pod, could you elaborate?`

I consider encapsulating the needed bits of a function in the same pod as a benefit, because it is easier (for me at least) to reason about what a `function` is, and how can I manage the workload, and compose/layering it independently of other functions.   It also makes less assumptions about things happening outside of the pod, and reduce the proxy overhead by having the proxy entity in the same pod.

phsiao commented:
`How would you use istio in front of a trigger side car? I thought, istio comes with Envoy as its own side car. Sorry, I don't know much about istio, yet. Could you describe, what you have in mind?`

@deissnerk I might be saying more than I know when it comes to istio.   But here is what I think how it could work.

Each runtime implements the interface, via http or other protocol.   The runtime only listens on a specific port on the localhost, for simplicity's sake.   Each trigger sidecar can then bridge the type of trigger they handle to the runtime via localhost communication.

istio, if I am not mistaken, can still be inserted here.   Take trigger-http as an example, istio will intercept the request to trigger-http sidecar like it normally would, and after processing put it back to the trigger-http sidecar.   istio, if I am not mistake, is using iptables to intercept requests, so it does not matter if it is a runtime that is handling the incoming requests directly, or it is a runtime.

Like I mentioned earlier, I can be completely wrong about how istio can fit in here.

andresmgot commented:
> these trigger-bridging sidecars should be small and take little resource

Even if they are small, I see it as a disadvantage. If the number of functions/replicas grow considerably it will be a significant amount of resources.

> ... and it is likely preferred because the interface of the function is different now ...

The interface of the function should be the same regardless of the trigger source.

> It also makes less assumptions about things happening outside of the pod

I see them completely detached. The runtime only cares about receiving HTTP request and replying to them. It doesn't need to make any assumptions as far as I can tell.

phsiao commented:
>> ... and it is likely preferred because the interface of the function is different now ...

> The interface of the function should be the same regardless of the trigger source.

Sorry, my use of "interface of the function" was not accurate, I meant to say how the rest of world interacts with the function has changed.  For example, you changed it from trigger-http to trigger-event, and now all the consumer of trigger-http is broken.

>> It also makes less assumptions about things happening outside of the pod
> I see them completely detached. The runtime only cares about receiving HTTP request and replying to them. It doesn't need to make any assumptions as far as I can tell.

The same can be said about sidecars, it just happens to co-locate in the same pod.

The "assumptions" that I am referring to are things like the ingress controller and kafka consumer in the diagram, that now we need to assume it is supported in a specific way.   I personally don't have a lot of experience with ingress controller, but I know there are many implementations and they all share/differentiate in some ways, and we need to make some assumptions about them.   And if we are spinning up a ingress controller/kafka consumer per function, and scale accordingly, would not we run into similar overhead at the end especially when the number of pods running the function is small?

deissnerk commented:
I think the question behind all the technical details is, if we regard triggers as independent entities, or if they are attributes of functions. E.g. in OpenWhisk triggers are independent entities. You create them independently from functions and use rules to connect functions and triggers. A trigger represents a certain event that can occur. In PubSub this can be compared with a topic. As a consequence a trigger could be defined for multiple functions. IMHO this idea of a trigger fits better to the implementation with separate pods, but it would not be a trigger pod per function. You would have at least one pod per trigger (i.e. event) defined in the system.
On the other hand, if a trigger is regarded as something tied to a specific function, the side car approach seems to be more natural to me. 

I personally prefer the idea of triggers as independent entities. You define what the events are that occur in your system and wire them to some logic. While playing around or debugging you can change the wiring. BTW would there be CRDs for triggers? 

@andresmgot @phsiao What do you think?

andresmgot commented:
I also think of triggers as an independent entity. I think we agree that we can differentiate three different logic pieces:
 - A source event controller (aka trigger controller). This entity would generically handle event sources. For example "Listen for messages in the topic `my-s3-bucket`".
 - A function (aka runtime). This entity just perform an action when called. For example: "Every time I receive a event I copy the files from one bucket to another"
 - A link between an event source and a function: Every time a message is received in the `my-s3-bucket` topic call the function `copy-to-other-bucket`.

In my opinion it feels natural to have the function and the trigger in different pods. It also makes sense to me to store relations between triggers and functions as CRDs. As I envision this, the trigger controller, once it detects an event, it should retrieve the links between events and functions and invoke the required functions with the event info.

phsiao commented:
@deissnerk @andresmgot Agreed, the way one treats triggers would impact the view of the implementation.   I personally think triggers are just bridge/glue that does very generic and simple transformation and invoke functions.   A function can have multiple triggers and it should just work fine without issues.   

But, I was not thinking about triggers should be added/changed at runtime, and this is one of the point that @andresmgot raised and it is one of the main point of this proposal.   I was thinking alone the line of triggers can't be changed once deployed with the function, but decouple the implementation of functions and triggers so they can be maintained separately with a clear interface between them.  My view is also that the triggers are part of the function interface as seen by the outside world, and if they are being changed (different topic, different schedule, different URL), the function has changed too and should be relaunched.

@deissnerk regarding to "a trigger pod per function", I think we will run into issue with permissions, security and routing if we don't bind the trigger pod at least to the namespace.

phsiao commented:
One more thoughts on the trigger being a separate entity.  Does that mean that kubeless now reduces the runtime to http trigger in today's form, and instead switches focus to authorizing, securing and routing requests between triggers and functions?

Sorry for all these questions, but I think it is helpful to have all these discussions.

phsiao commented:
>> Does that mean that kubeless now reduces the runtime to http trigger in today's form, and instead switches focus to authorizing, securing and routing requests between triggers and functions?

> I think I don't fully understand that question. The runtime (function container) would receive an HTTP request in any case (containing all the information about the event source, probably using custom headers). Then, the runtime, before giving the control to the function code, will transform the HTTP request to the interface format that the document details. Is that what you meant?

Was not clear to me if it is a trigger pod per namespace or not, assuming it is not, then:

My concern is then you will have trigger pod living in some namespace, while functions leaving in some others.   How does a function restrict access to its http port if it needs to?  We can leverage NetworkPolicy but it is not required to have it to run a cluster.

Also routing, if they are in different namespaces, how do you find the function runtimes?  I assume that will be using the endpoints objects, but the trigger may not have permissions to look that up.   The controller can help bridge it, but that does not seem very natural to me.

> Right now we will not focus in improving authorization or security.

We may need to handle it depending on the architecture that we decided.

andresmgot commented:
> Was not clear to me if it is a trigger pod per namespace or not, assuming it is not, then:

I thought on having just one pod (in the `kubeless` namespace). 

> How does a function restrict access to its http port if it needs to? We can leverage NetworkPolicy but it is not required to have it to run a cluster

For the moment I think we are not going to allow that. We can improve the security capabilities in another workflow. (As far as it is doable it is fine for now)

> Also routing, if they are in different namespaces, how do you find the function runtimes? I assume that will be using the endpoints objects, but the trigger may not have permissions to look that up. The controller can help bridge it, but that does not seem very natural to me.

If necessary, we will discover function runtimes using labels. We don't need to use the controller to retrieve the running functions if we have the information stored as CRDs or anything similar. We can store there, along with the function information, the namespace in which they are running. The trigger should have permissions enough to reach those pods.

phsiao commented:
In that case, I can see twl limitations that I personally don't agree

1. trigger-http would have a discoverable name that is outside of the namespace that it is created, essentially the ingress controller's service name.  That also reduces the function of namespace to be just resource container and not for permission/security context.

2. Also for trigger-http, function name now needs to be unique, unless the namespace is going to be encoded in in the alias for ingress controller, but I am not sure how that is going to work.   And this can be bad, because a function from different namespace can take over the function from another namespace, but produces very different function.

On the topic of security, I would encourage us to at least plan it out ahead of time by describing how that can be supported.  Often time security is bolted on after the fact and that usually don't work very well.

Sharing trigger also has the issue that what if the trigger needs function specific credentials, such as auth for kafka and other similar situation, where do we store that information?

andresmgot commented:
Those are fair points. We may not be able to do namespace isolation (at least with the default version of the Ingress controller).

Regarding the issue that the function name needs to be unique regardless of the namespace, it is something that also exists now, maybe we should tackle it in a different issue?

> Sharing trigger also has the issue that what if the trigger needs function specific credentials, such as auth for kafka and other similar situation, where do we store that information?

I don't think that will be a problem. In that case we could add the authentication information in Secrets and link to them in the the object that stores the link between the trigger and the runtime (probably a CRD).

Let's spend some time investigating those issues before taking a decision.

andresmgot commented:
@phsiao I finally have time to re-think about this. Let me try to elaborate my take about your concerns (assuming that we would have just one trigger pod per kubeless deployment):

> trigger-http would have a discoverable name that is outside of the namespace that it is created, essentially the ingress controller's service name. 

As far as I know, the name of the http-trigger doesn't need to be discoverable. It should be reachable by the function pod though. The trigger-http would work just as a proxy between the edge router and the runtime container. In case we want authentication we could use Istio, but for the first version, a simple Ingress Controller could do the job. As far as the ingress/egress traffic is properly controlled by a network policy (if needed) I don't see a security problem there.

> That also reduces the function of namespace to be just resource container and not for permission/security context.

That can be controlled as well with network policies. It is true that if the http-trigger is in the same pod any contact with the `kubeless` namespace would be needed but as far as we set proper policies, it should be a problem to open the access to the http-trigger. The same applies to the Kafka trigger. 

> Also for trigger-http, function name now needs to be unique, unless the namespace is going to be encoded in in the alias for ingress controller, but I am not sure how that is going to work.

The function definition is just present in the function namespace (you can have several functions with the same ID in different namespaces). The same goes for the ConfigMap, Service, Deployment... all of them are namespace so it just requires to be unique per namespace. The same goes for the Ingress rule, they are namespaced so we just need to provide the pods a hostname including the namespace. This could be an example:

```
  apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    name: ingress
    namespace: dev-1
  spec:
    rules:
    - host: get-python.default.com
      http:
        paths:
        - backend:
            serviceName: get-python
            servicePort: 8080
          path: /
```

That way the only thing that needs to be unique is the rule host. The same goes for Kafka, each consumer will have on its `group ID` the function namespace so functions along different namespaces can use the same function ID and still receive messages.

> And this can be bad, because a function from different namespace can take over the function from another namespace, but produces very different function.

That is protected by design, an Ingress rule from one namespace cannot access a service from a different namespace.

As far as I can see, the proposed approach won't cause a major issue whenever we tackle the multi tenancy issue. Let me know if you still think otherwise.

phsiao commented:
@andresmgot and I had a conversation offline on this, and it comes down to my concerns or "bias" towards adding another pod between caller and function.  In clusters where pod IP is routable from outside, this can introduce a undesirable overhead that does not have a lot of value.  But I also recognize that routable pod IP is probably not a norm.

andresmgot commented:
the conversation is available in the Bitnami OSS slack channel: https://bitnami-oss.slack.com/archives/C5A89A32T/p1510673021000421

murali-reddy commented:
Joining late to the thread. Sorry if the questions were discussed earlier or on other issues.

Wondering if we should have an universal script for all event triggers. For instance don't see a reason why we can not merge [event-trigger](https://github.com/kubeless/kubeless/blob/master/docker/runtime/python-2.7/event-trigger/kubeless.py) and [http-trigger](https://github.com/kubeless/kubeless/blob/master/docker/runtime/python-2.7/http-trigger/kubeless.py) into a single script. Depending on the argument we can have event trigger or http trigger setup.

Pushing it further. If we assume python is available on all run times, or use golang static binary then we dont need to write language/runtime specfic triggers as well.

Ofcourse we will need a language agnostic way to invoke language specific function. Has any thought was given to use gRPC? Bascially run a gRPC server by using language specific gRPC wrapper for the function, so event trigger as gRPC client can invoke function.

/cc @sebgoa @anguslees

andresmgot commented:
> Wondering if we should have an universal script for all event triggers. For instance don't see a reason why we can not merge event-trigger and http-trigger into a single script. Depending on the argument we can have event trigger or http trigger setup.

What do you mean with "single script"? I guess you mean a single processor, that's the idea, having a script/wrapper that receives an HTTP request with the information about the event source and pass it to the function code. Then it can return the response to the event source. That's what I wanted to expose in the diagram of the PR.

> Pushing it further. If we assume python is available on all run times, or use golang static binary then we dont need to write language/runtime specfic triggers as well.

Yes, that's the goal, implement just one trigger (probably in Go) and reuse it for calling any language/runtime.

> Ofcourse we will need a language agnostic way to invoke language specific function. Has any thought was given to use gRPC? Bascially run a gRPC server by using language specific gRPC wrapper for the function, so event trigger as gRPC client can invoke function.

We evaluated gRPC but it is more complex that HTTP so it would make it more difficult to develop new trigger sources/runtimes. For the moment we decide to go with HTTP.

andresmgot commented:
I am taking a look at how nuclio does this and I like the concept (https://github.com/nuclio/nuclio/tree/8d7a9dfa21785fbd26f4654d75f090af53bc6021/pkg/processor/trigger). 

They have a `trigger` interface that implement the required methods (start, stop, process event). They implement the interface for several event sources and basically what they do is connect to the event source (rabbitmq, kafka, nats...) and whenever a message is received they redirect the request to a local daemon that is the one that run the function (python, golang). 

This doesn't require nor an additional pod or a sidecar container since they will have several daemons (processes) running in the function container:
 - The native runtime daemon: This is the web server written in the target runtime (python, ruby, javascript...) that receives a request, execute the user function and returns a response.
 - The event source listener(s): They have as many processes of this type as event sources are configured. Each fo this listener redirects messages from the source to the native runtime daemon.
 - The orchestrator (called `processor`): This is the main (PID 1) process. Basically it reads the configuration and start and many processes as needed. It also handles other resources like logging or binding storage.

This has the disadvantage that the `processor` should monitor all its subprocesses so if any of them dies it should restart it (nuclio doesn't handle that right now). Also for each change in the trigger configuration it should redeploy the function.

andresmgot commented:
So since we we agree in the interface format for the function I will start working on that. Let's try to divide the problem in small ones.

deissnerk commented:
Maybe, we should have a look at the original design goals:
> We should design a runtime abstraction. So that:
> - Triggers can be added (http or event or something else).
> - These can be in a single language i.e golang
> - Runtimes can be added more easily
> - One function can be triggered by more than one trigger source

To the last one in the list I would also like to add that multiple functions might be triggered by a single trigger. To me a trigger really is an entity on its own. This of course does not necessarily rule out  a sidecar approach. We could still have something similar to what istio does with envoy, but while envoy really wraps the network I/O of a pod, a 1:1 relation of trigger and function runtime would be artificial.

I think, there may also be a few more requirements to consider:
- Functions need to be controlled by the serverless framework. It should be possible to kill them, e.g. on timeout.
- Kubeless might be used by someone to provide a service to a group of users. In this case it might be desirable to bill the users per function execution as it is known from existing offerings.

As a service provider you need to consider that a user might try to manipulate the serverless environment. I would prefer to isolate the actual function as much as possible from the framework. 

A trigger in a separate pod can do much better error handling, if a function pod fails or times out. It can then try to call another function pod or execute some other error handling function. You could even think of function failures as triggers for other functions. :-)

deissnerk commented on docs/proposals/decoupling-triggers-and-runtimes.md:
```diff
@@ -0,0 +1,117 @@
+
+ - Data: Serialized version of the request body
+ - Metadata: Information about the request
+   - Type: Type of the request ("HTTP", "PubSub", "Scheduled", ...)
```
I propose to adopt the 4 invocation types that were mentioned in the whitepaper of serverless-wg of CNCF:
- synchronous
- asynchronous
- streaming
- batch

E.g. HTTP could be both. Just think of a web hook. In the asynchronous HTTP case, the response could be just a tracking ID and not the result of the function.

anguslees replied:
I strongly believe triggers should be dumb, and cleverness should be in (composable) functions.

So for example your async HTTP example is just a _sync_ HTTP trigger+function that puts an event on a queue and then returns a tracking ID (again, synchronously).  This pattern can be used to turn any sync request into async (and you can think of the opposite version that blocks the caller until some sort of "async" response occurs).

I suggest *triggers* are inherently sync or async (or other properties you mention).  The differences are the way the *function* is invoked by that trigger - specifically, if it's worth generating a response object.  In this light, _all_ functions could be invoked synchronously, and we could just document that (eg) pubsub triggers discard any return value.

More interesting to me is "at most once" vs "at least once".  In particular: should the caller (ie: trigger) re-attempt calling the function several times until it gets a response, or should it "fire and forget" and return no response (not the usual interpretation of at-most-once, but useful in this context).  The at-most-once/at-least-once concept can be applied obviously to pubsub queues (using the same terminology), _and_ many types of HTTP notification/monitoring webhooks when we don't care about the response (even failure) (this is sometimes called level-triggered vs edge-triggered).

andresmgot replied:
Let's omit the `type` key, we can substitute it with a `source` so functions can adapt the logic, if necessary, depending on the source.

Regarding sync/async I agree to make functions always "sync". The returned value of the function can be "stringified" and sent back to the caller. That way functions can return function final results or a track ID.

Finally regarding the behavior, let's go with "at most once" since it is more similar to the way we are doing it today. If in the future we want to add the possibility of configuring a trigger with the "at least once" policy we will be able to do it.

deissnerk commented on docs/proposals/decoupling-triggers-and-runtimes.md:
```diff
@@ -0,0 +1,117 @@
+So far we can identify two types of trigger, we will call them _trigger controllers_:
+ - HTTP Trigger: This trigger should redirect HTTP(s) requests from/to the _runtime_. For doing so we can use an Ingress Controller. That way, we can translate each _HTTP Trigger_ for an Ingress Rule.
+ - Kafka Topic Trigger: This trigger should translate topic messages to HTTP request. This way runtimes can have an unique interface, regardless of its trigger. Each _kafka trigger_ can be translated to a Kafka consumer.
+
```
I would think of trigger controllers as actual k8s controllers, as triggers should be represented by custom resources on their own. The trigger controllers could then listen for them and create an ingress or a kafka message consumer or whatwever else is needed.

**Open questions to me would be:**
- Should each trigger type have its own CRD, or should there be a common CRD? The different trigger types could still get their own annotations for the controller to look for.
- Should there be a custom resource for each event (e.g. a topic, an HTTP endpoint etc.) and for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.

murali-reddy replied:
>I would think of trigger controllers as actual k8s controllers, as triggers should be represented by custom resources on their own.

+1. IMO, Function, Trigger and Runtime should be the first class citizen/entity of Kubeless. Each of them should have a CRD. Invoication type, topic name etc can be attribute of the Trigger.

>for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.

+1 for rule type abstraction to associate Function and Trigger. This would also enable n:m association between function and triggers.

Some thing to note though is these changes will be breaking changes.

joek replied:
> for each event-function association or trigger? At least from OpenWhisk I know that they have triggers and rules.
One Comment here, as we might have to treat things different. On OpenWhisk they have an API Gateway to expose functions via http. The API Gateway creates an 1:1 realtion between the function and the endpoint (no rules are used here). Triggers are used for more high level async integrations like messaging or chat applications. 

From my point of view, it makes sense to have a trigger + rules concept for async processing. For sync processing (using http) I would like to keep it as simple as possible and reduce the complexity to reduce the response time.

andresmgot replied:
For the moment let's take the simple approach: A single CRD per trigger. Each instance of this CRD will contain the references for the functions it should trigger. The trigger controller would create then Ingress rules, Kubernetes services, Kafka consumers or whatever is needed in order to call the function with the less latency possible.

andresmgot commented:
@deissnerk @anguslees @joek @murali-reddy I have updated the proposal. I have included everyone comments and I have simplified it to not enter into implementation details around trigger CRDs and controllers (so they can be handled in a different PR and close this one)

Please let me now if you have any additional comment. We will start with the implementation in the next days. Thank you for your patience and feedback!

andresmgot commented:
Thank you everyone for your feedback! We are merging this now and start working in the implementation.
